packages/xo-server/src/xapi/mixins/patching.mjs

```diff
@@ -587,16 +603,117 @@ const methods = {
     throw new Error('non pool-wide install not implemented')
   },
 
-  async rollingPoolUpdate($defer, parentTask, { xsCredentials } = {}) {
+  async xsCdnUpdate(hosts, hash) {
+    if (hosts === undefined) {
+      hosts = Object.values(this.objects.indexes.type.host)
+    }
+
+    if (hash === undefined) {
+      hash = (await this._fetchXsUpdatesEndpoint(hosts[0])).hash
+    }
+
+    // Hosts need to be updated one at a time starting with the pool master
+    hosts = hosts.sort(({ $ref }) => ($ref === this.pool.master ? -1 : 1))
+
+    for (const host of hosts) {
+      await this.callAsync('host.apply_updates', host.$ref, hash)
+    }
+
+    await this._fetchXsUpdatesEndpoint(REMOVE_CACHE_ENTRY, hosts[0])
+  },
+
+  async _getPendingGuidances(object, level = PENDING_GUIDANCES_LEVEL.full) {
+    const record = await this.getRecord(object.$type, object.$ref)
+    const pendingGuidances = new Set()
+
+    if (record.pending_guidances.length >= PENDING_GUIDANCES_LEVEL.mandatory && record.pending_guidances.length > 0) {
+      pendingGuidances.add(...record.pending_guidances)
+    }
+
+    if (level >= PENDING_GUIDANCES_LEVEL.recommended && record.pending_guidances_recommended.length > 0) {
+      pendingGuidances.add(...record.pending_guidances_recommended)
+    }
+    if (level >= PENDING_GUIDANCES_LEVEL.full && record.pending_guidances_full.length > 0) {
+      pendingGuidances.add(...record.pending_guidances_full)
+    }
+
+    return Array.from(pendingGuidances)
+  },
+
+  async _pendingGuidancesGuard(object, level) {
+    const pendingGuidances = await this._getPendingGuidances(object, level)
+
+    if (pendingGuidances.length > 0) {
+      /* throw */ incorrectState({
+        actual: pendingGuidances,
+        expected: [],
+        object: object.uuid,
+        property: 'pending_guidances(_recommended|_full)',
+      })
+    }
+  },
+
+  // for now, only handle VM's pending guidances
+  async _applyPendingGuidances(vm, pendingGuidances) {
+    if (pendingGuidances.length === 0) {
+      return
+    }
+
+    if (pendingGuidances.includes('restart_vm')) {
+      return this.VM_reboot(vm.$ref, { bypassBlockedOperation: true })
+    }
+
+    if (pendingGuidances.includes('restart_device_model')) {
+      try {
+        return await this.callAsync('VM.restart_device_models', vm.$ref)
+      } catch (error) {
+        log.debug(`restart_device_models failed on ${vm.uuid}. Going to reboot the VM`, error)
+        return this.VM_reboot(vm.$ref, { bypassBlockedOperation: true })
+      }
+    }
+
+    throw new Error(`Pending guidances not implemented: ${pendingGuidances.join(',')}`)
```

Comment:
is it possible to use notImplemented error ? to be coherent with the incorrectstate